### ISSUE TITLE
在Macos sierra beta2上编译bilibili，卡在最后一步，请高手指点迷津！

### ISSUE THREAD

**Issue by xuejunli** [typcn/bilibili-mac-client#293]
系统环境：
Macos sierra beta2
git
xcode 8 beta2

按照作者的指导文件，先下载源码，发现FMDB文件夹是空的，编译，报错，找不到FMDB.h，到Ｇithub上下载了FMDB系列\chromium-tabs系列＼danmaku2ass系列＼GCDWebServer系列，编译的时候再不报任何找不到文件的错误了；
但最后一步：
ld: framework not found FMDB
clang: error: linker command failed with exit code 1 (use -v to see invocation)

实在走不下去了，请各位高手不吝赐教！

**Comment by typcn** [comment]
因为你没有按照编译说明
https://github.com/typcn/bilibili-mac-client/blob/master/HOW_TO_BUILD.md

**Comment by xuejunli** [comment]
我是按照这个文档操作的，先安装homebrew，然后安装git，然后用git clone源码，不过我发现 git submodule update --init 运行后没有任何作用，理解这个应该是把引用的公用代码库下载下来的。

**Comment by typcn** [comment]
再运行下 git submodule update 试试？

**Comment by xuejunli** [comment]
git submodule update 一闪而过，没有任何变化

在线支持啊，感谢！

**Comment by xuejunli** [comment]
我把Ａpple llvm 8.0- language- c++的两个编译选项从Ｃ＋＋11 变更为compiler default后重新编译，就只有两个报错了：

public:
    tcp_client();
    bool conn(std::string, int, void *userdata);
    bool send_data(const void *data,int size);
    std::string receive(int);
    void disconnect😈();    # //Non-Ascii characters are not allowed outside of literals and identifiers
};

**Comment by typcn** [comment]
。。。。。。。。。。。。。。

**Comment by typcn** [comment]
你把 repo 删了重新 clone 然后再  init submodule 试试

**Comment by xuejunli** [comment]
稍等，我再按照您的删除再来，目前我根据提示错误自动修正了一下，只在Ａirplay下面还有一个错误：
    string result = rhttp->receive(2048);
    if(result.find("101") != string::npos){
        std::thread t1(revReply,rhttp);  // No matching constructor for initialization of 'std::thread'
        t1.detach();
        return true;
    }else{
        return false;
    }
}

**Comment by typcn** [comment]
你改成非 C++ 11 ，只是暴露出用了新特性的两个语法错误，你修完了语法错误，之前那些错误才会出现。。。

**Comment by typcn** [comment]
语法错误 > 编译错误 > 链接错误
每一部分都是独立的，一次显示的编译错误数量还跟编译进程数/编译器设置有关

**Comment by xuejunli** [comment]
已经删除，git bilibili了，到该目录下，正在git submodule --init，我把library目录下之前生成的临时文件都删除了

**Comment by xuejunli** [comment]
重新下载了，默认编译还是 16个错，有没有什么方法你来远程控制我的电脑来试试看，我用的macos Ｓierra beta2？

**Comment by typcn** [comment]
截图

**Comment by typcn** [comment]
如果 mpv 什么的没链接上的话，那你一定是没安装 git-lfs，或者下载服务器被墙导致失败了

可以手动从这个文件夹下载
https://github.com/typcn/bilibili-mac-client/tree/master/bilibili/libs

**Comment by typcn** [comment]
或者你可以从我官网的 APP 里面，右键，查看包内容，Contents/MacOS/libs ，把里面的复制到源码目录 bilibili/libs 并覆盖

**Comment by xuejunli** [comment]
好，按照你的方法再来一次，看http://blog.csdn.net/zuoyou1314/article/details/41678527 也是说少了架构和库

**Comment by xuejunli** [comment]
复制了库文件以后，果然编译成功，不过编译成功文件还是不能正常运行。

请教一下编译成功的文件怎么运行才行？

**Comment by typcn** [comment]
直接点击Xcode左上角的运行就可以运行了

**Comment by xuejunli** [comment]
我直接在xcode中运行是可以的，只是有提示而已！

**Comment by typcn** [comment]
是的。。就在 Xcode 里面运行吧。。苹果的 BUG。。。。10.12 上复制出来运行不了

**Comment by typcn** [comment]
太傻逼了，DP1 刚出来我就去给苹果发了将近十个报告了，DP1 的时候 Xcode 编译复制出来还能运行，现在系统出了新版，苹果不但一个都没修复，TMD 反而问题更多了。。。。。。

**Comment by xuejunli** [comment]
好的，谢谢！
我是菜鸟，添麻烦了！

**Comment by xuejunli** [comment]
xcode下勉强能运行，实际上也做不了什么，比如点击播放或者搜索都会崩溃：
异常名称：NSRangeException
描述：*** -[__NSArray0 objectAtIndex:]: index 0 beyond bounds for empty NSArray
用户信息：(null)

0   CoreFoundation                      0x00007fff7cb921cb __exceptionPreprocess + 171
1   libobjc.A.dylib                     0x00007fff906378fa objc_exception_throw + 48
2   CoreFoundation                      0x00007fff7cbac5cd -[__NSArray0 objectAtIndex:] + 93
3   Bilibili                            0x000000010002cd0c -[WebTabView viewFrameDidChange:] + 460
4   ChromiumTabs                        0x0000000100dbe533 -[CTTabContentsController ensureContentsVisible] + 403
5   ChromiumTabs                        0x0000000100d9eee0 -[CTTabStripController swapInTabAtIndex:] + 512
6   ChromiumTabs                        0x0000000100da80d0 -[CTTabStripController tabSelectedWithContents:previousContents:atIndex:userGesture:] + 1072
7   ChromiumTabs                        0x0000000100da7c46 -[CTTabStripController tabDidSelect:] + 374
8   CoreFoundation                      0x00007fff7cb2028c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
9   CoreFoundation                      0x00007fff7cb2018b _CFXRegistrationPost + 427
10  CoreFoundation                      0x00007fff7cb1fef2 ___CFXNotificationPost_block_invoke + 50
11  CoreFoundation                      0x00007fff7cadd412 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 2018
12  CoreFoundation                      0x00007fff7cadc3fb _CFXNotificationPost + 667
13  Foundation                          0x00007fff7e51dd07 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
14  ChromiumTabs                        0x0000000100dae2f8 -[CTTabStripModel changeSelectedContentsFrom:toIndex:userGesture:] + 664
15  ChromiumTabs                        0x0000000100dab05e -[CTTabStripModel insertTabContents:atIndex:withAddTypes:] + 846
16  ChromiumTabs                        0x0000000100dacdf9 -[CTTabStripModel addTabContents:atIndex:withTransition:addTypes:] + 409
17  ChromiumTabs                        0x0000000100dafdc7 -[CTBrowser addTabContents:atIndex:inForeground:] + 151
18  ChromiumTabs                        0x0000000100dafe79 -[CTBrowser addTabContents:inForeground:] + 89
19  Bilibili                            0x0000000100021e83 -[TWebView webView:createWebViewWithConfiguration:forNavigationAction:windowFeatures:] + 595
20  WebKit                              0x00007fff8482075c _ZN6WebKit10UIDelegate8UIClient13createNewPageEPNS_12WebPageProxyEPNS_13WebFrameProxyERKN7WebCore18SecurityOriginDataERKNS6_15ResourceRequestERKNS6_14WindowFeaturesERKNS_20NavigationActionDataE + 430
21  WebKit                              0x00007fff848b3bbc _ZN6WebKit12WebPageProxy13createNewPageEyRKN7WebCore18SecurityOriginDataERKNS1_15ResourceRequestERKNS1_14WindowFeaturesERKNS_20NavigationActionDataERyRNS_25WebPageCreationParametersE + 148
22  WebKit                              0x00007fff848cad2a _ZN3IPC13handleMessageIN8Messages12WebPageProxy13CreateNewPageEN6WebKit12WebPageProxyEMS5_FvyRKN7WebCore18SecurityOriginDataERKNS6_15ResourceRequestERKNS6_14WindowFeaturesERKNS4_20NavigationActionDataERyRNS4_25WebPageCreationParametersEEEEvRNS_14MessageDecoderERNS_14MessageEncoderEPT0_T1_ + 502
23  WebKit                              0x00007fff847872fd _ZN3IPC18MessageReceiverMap19dispatchSyncMessageERNS_10ConnectionERNS_14MessageDecoderERNSt3__110unique_ptrINS_14MessageEncoderENS5_14default_deleteIS7_EEEE + 143
24  WebKit                              0x00007fff84904464 _ZN6WebKit15WebProcessProxy21didReceiveSyncMessageERN3IPC10ConnectionERNS1_14MessageDecoderERNSt3__110unique_ptrINS1_14MessageEncoderENS6_14default_deleteIS8_EEEE + 28
25  WebKit                              0x00007fff84756490 _ZN3IPC10Connection19dispatchSyncMessageERNS_14MessageDecoderE + 198
26  WebKit                              0x00007fff84753cf6 _ZN3IPC10Connection15dispatchMessageENSt3__110unique_ptrINS_14MessageDecoderENS1_14default_deleteIS3_EEEE + 104
27  WebKit                              0x00007fff84756650 _ZN3IPC10Connection18dispatchOneMessageEv + 126
28  JavaScriptCore                      0x00007fff7f924f69 _ZN3WTF7RunLoop11performWorkEv + 169
29  JavaScriptCore                      0x00007fff7f925182 _ZN3WTF7RunLoop11performWorkEPv + 34
30  CoreFoundation                      0x00007fff7cb2a111 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
31  CoreFoundation                      0x00007fff7cb0a4cc __CFRunLoopDoSources0 + 556
32  CoreFoundation                      0x00007fff7cb099b6 __CFRunLoopRun + 934
33  CoreFoundation                      0x00007fff7cb093ad CFRunLoopRunSpecific + 285
34  HIToolbox                           0x00007fff7c0c8bcc RunCurrentEventLoopInMode + 240
35  HIToolbox                           0x00007fff7c0c8a01 ReceiveNextEventCommon + 432
36  HIToolbox                           0x00007fff7c0c8836 _BlockUntilNextEventMatchingListInModeWithFilter + 71
37  AppKit                              0x00007fff7a83ca91 _DPSNextEvent + 1093
38  AppKit                              0x00007fff7af46704 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1658
39  AppKit                              0x00007fff7a831409 -[NSApplication run] + 926
40  AppKit                              0x00007fff7a7fcaea NSApplicationMain + 1237
41  Bilibili                            0x000000010001cf92 main + 34
42  libdyld.dylib                       0x00007fff90f10255 start + 1
43  ???                                 0x0000000000000003 0x0 + 3

还是等sierra自己解决问题吧。

**Comment by typcn** [comment]
WebTabView viewFrameDidChange
这个地方的崩溃，是我读取 webview 的 subviews ，然后他就崩溃了
也是苹果的 BUG。。。。我也是服。。。本来跑的好好的改出这么多 BUG 来

**Comment by xuejunli** [comment]
beta3之后编译还是一样的效果，等.......

**Comment by typcn** [comment]
.... 苹果这是不准备修了么

**Comment by xuejunli** [comment]
beta4发布了，重新下载了源码编译，编译成功，在编译环境可以运行，点击播放报错，和之前一样：
异常名称：NSRangeException
描述：*** -[__NSArray0 objectAtIndex:]: index 0 beyond bounds for empty NSArray
用户信息：(null)

0   CoreFoundation                      0x00007fffca2ef5eb __exceptionPreprocess + 171
1   libobjc.A.dylib                     0x00007fffdeb659ba objc_exception_throw + 48
2   CoreFoundation                      0x00007fffca30978d -[__NSArray0 objectAtIndex:] + 93
3   Bilibili                            0x000000010002b53c -[WebTabView viewFrameDidChange:] + 460
4   ChromiumTabs                        0x0000000100db9533 -[CTTabContentsController ensureContentsVisible] + 403
5   ChromiumTabs                        0x0000000100d99e80 -[CTTabStripController swapInTabAtIndex:] + 512
6   ChromiumTabs                        0x0000000100da30d0 -[CTTabStripController tabSelectedWithContents:previousContents:atIndex:userGesture:] + 1072
7   ChromiumTabs                        0x0000000100da2c46 -[CTTabStripController tabDidSelect:] + 374
8   CoreFoundation                      0x00007fffca27c51c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
9   CoreFoundation                      0x00007fffca27c41b _CFXRegistrationPost + 427
10  CoreFoundation                      0x00007fffca27c182 ___CFXNotificationPost_block_invoke + 50
11  CoreFoundation                      0x00007fffca23a8f2 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 2018
12  CoreFoundation                      0x00007fffca2398db _CFXNotificationPost + 667
13  Foundation                          0x00007fffcbc95fd3 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
14  ChromiumTabs                        0x0000000100da92f8 -[CTTabStripModel changeSelectedContentsFrom:toIndex:userGesture:] + 664
15  ChromiumTabs                        0x0000000100da605e -[CTTabStripModel insertTabContents:atIndex:withAddTypes:] + 846
16  ChromiumTabs                        0x0000000100da7df9 -[CTTabStripModel addTabContents:atIndex:withTransition:addTypes:] + 409
17  ChromiumTabs                        0x0000000100daadc7 -[CTBrowser addTabContents:atIndex:inForeground:] + 151
18  ChromiumTabs                        0x0000000100daae79 -[CTBrowser addTabContents:inForeground:] + 89
19  Bilibili                            0x00000001000205e3 -[TWebView webView:createWebViewWithConfiguration:forNavigationAction:windowFeatures:] + 595
20  WebKit                              0x00007fffd2739896 _ZN6WebKit10UIDelegate8UIClient13createNewPageEPNS_12WebPageProxyEPNS_13WebFrameProxyERKN7WebCore18SecurityOriginDataERKNS6_15ResourceRequestERKNS6_14WindowFeaturesERKNS_20NavigationActionDataE + 536
21  WebKit                              0x00007fffd27cae54 _ZN6WebKit12WebPageProxy13createNewPageEyRKN7WebCore18SecurityOriginDataERKNS1_15ResourceRequestERKNS1_14WindowFeaturesERKNS_20NavigationActionDataERyRNS_25WebPageCreationParametersE + 148
22  WebKit                              0x00007fffd27e298b _ZN3IPC13handleMessageIN8Messages12WebPageProxy13CreateNewPageEN6WebKit12WebPageProxyEMS5_FvyRKN7WebCore18SecurityOriginDataERKNS6_15ResourceRequestERKNS6_14WindowFeaturesERKNS4_20NavigationActionDataERyRNS4_25WebPageCreationParametersEEEEvRNS_14MessageDecoderERNS_14MessageEncoderEPT0_T1_ + 499
23  WebKit                              0x00007fffd269d103 _ZN3IPC18MessageReceiverMap19dispatchSyncMessageERNS_10ConnectionERNS_14MessageDecoderERNSt3__110unique_ptrINS_14MessageEncoderENS5_14default_deleteIS7_EEEE + 143
24  WebKit                              0x00007fffd281c9f8 _ZN6WebKit15WebProcessProxy21didReceiveSyncMessageERN3IPC10ConnectionERNS1_14MessageDecoderERNSt3__110unique_ptrINS1_14MessageEncoderENS6_14default_deleteIS8_EEEE + 28
25  WebKit                              0x00007fffd266be90 _ZN3IPC10Connection19dispatchSyncMessageERNS_14MessageDecoderE + 198
26  WebKit                              0x00007fffd26696f6 _ZN3IPC10Connection15dispatchMessageENSt3__110unique_ptrINS_14MessageDecoderENS1_14default_deleteIS3_EEEE + 104
27  WebKit                              0x00007fffd266c050 _ZN3IPC10Connection18dispatchOneMessageEv + 126
28  JavaScriptCore                      0x00007fffcd0f45a9 _ZN3WTF7RunLoop11performWorkEv + 169
29  JavaScriptCore                      0x00007fffcd0f47c2 _ZN3WTF7RunLoop11performWorkEPv + 34
30  CoreFoundation                      0x00007fffca2863f1 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
31  CoreFoundation                      0x00007fffca26784c __CFRunLoopDoSources0 + 556
32  CoreFoundation                      0x00007fffca266d36 __CFRunLoopRun + 934
33  CoreFoundation                      0x00007fffca266734 CFRunLoopRunSpecific + 420
34  HIToolbox                           0x00007fffc980df6c RunCurrentEventLoopInMode + 240
35  HIToolbox                           0x00007fffc980dda1 ReceiveNextEventCommon + 432
36  HIToolbox                           0x00007fffc980dbd6 _BlockUntilNextEventMatchingListInModeWithFilter + 71
37  AppKit                              0x00007fffc7f0d0b5 _DPSNextEvent + 1093
38  AppKit                              0x00007fffc861c6bf -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1637
39  AppKit                              0x00007fffc7f01a7c -[NSApplication run] + 926
40  AppKit                              0x00007fffc7ecc5d6 NSApplicationMain + 1237
41  Bilibili                            0x000000010001b6f2 main + 34
42  libdyld.dylib                       0x00007fffdf43b255 start + 1
43  ???                                 0x0000000000000003 0x0 + 3

**Comment by typcn** [comment]
妈的，webview 那个地方，我调试器段点下来显示有一个 subview，但是你读取一下直接越界出错...........

**Comment by typcn** [comment]
数组长度是1，读取0的地方异常崩溃....

**Comment by xuejunli** [comment]
网上找的资料：
http://stackoverflow.com/questions/31622530/nsarray-objectatindex-index-0-beyond-bounds-for-empty-array

http://blog.csdn.net/zyx196/article/details/49493029

**Comment by typcn** [comment]
不....挂上调试器可以看到这个数组是有成员的....

**Comment by typcn** [comment]
这个其实可以做一个 workaround
https://github.com/typcn/bilibili-mac-client/blob/8e51eb5064a66f36150783c18e4073101a1abfcb/bilibili/Browser/WebTabView.mm#L49
这一行，改成 HudView = [[NSView alloc] init]

**Comment by xuejunli** [comment]
修改后就ok了，谢谢，您可以直接发布兼容10.12的新版本了！

**Comment by typcn** [comment]
2.45 已兼容
